### PR TITLE
Feat/3049  share TLM connection between tabs using a broadcast channel

### PR DESCRIPTION
**Pull request by raphj**
#3049 

## Checkpoints

These points must be checked before merging. Please don't edit them out.

**For developers**

- [X] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [X] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [X] Automated tests covering the feature have **NOT** been written: this seems hard to test automatically, this would require open multiple table in cypress or something like this

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [ ] Manual, quality tests have been done

### CODE REVIEW

File: frontend/src/util/LiveMessageManager.js [hunk 1]
```diff
@@ -9,19 +10,24 @@ export const LIVE_MESSAGE_STATUS = {
   HEARTBEAT_FAILED: 'heartbeat_failed'
 }
 
+// increment this number each time the channel protocol is changed in an incompatible way
```
Comment (gbonaspetti) [inline frontend/src/util/LiveMessageManager.js:13]: This comment does not follow the pattern:
// INFO - RJ - 2020-08-12  - increment this number each time the channel protocol is changed in an incompatible way

File: frontend/src/util/LiveMessageManager.js [hunk 2]
```diff
@@ -126,10 +160,19 @@ export class LiveMessageManager {
...
   handleHeartbeatFailure () {
-    this.setStatus(LIVE_MESSAGE_STATUS.HEARTBEAT_FAILED)
+    this.broadcastAndSetStatus(LIVE_MESSAGE_STATUS.HEARTBEAT_FAILED)
     this.restartLiveMessageConnection()
   }
 
+  broadcastAndSetStatus (status) {
+    this.setStatus(status)
+    this.broadcastStatus(status)
+  }
```
[the remark this answers is not in the document]
Comment (raphj) [reply frontend/src/util/LiveMessageManager.js]: it won't, in fact.

File: frontend/src/container/Header.jsx [hunk 3]
```diff
@@ -77,7 +77,7 @@ export class Header extends React.Component {
 
     const fetchPostUserLogout = await props.dispatch(postUserLogout())
     if (fetchPostUserLogout.status === 204) {
-      props.tlm.manager.closeLiveMessageConnection()
+      LiveMessageManager.getInstance().closeLiveMessageConnection()
```
Comment (grignards) [inline frontend/src/container/Header.jsx:80]: I'm not sure that introducing a singleton is a really good idea as it makes the code more difficult to test and restrict the used class to have only one instance (which is currently what it is, but we cannot totally be sure it will always be the case). See https://www.michaelsafyan.com/tech/design/patterns/singleton for an argumentation against singletons.

Did you d this to get rid of un-necessary refreshes of the application? If yes, couldn't we find a way to keep the object passed as props without those refreshes?

File: frontend/src/util/LiveMessageManager.js [hunk 4]
```diff
@@ -70,7 +73,11 @@ export class LiveMessageManager {
     this.closeEventSourceConnection()
 
     this.eventSource = new EventSource(
-      `${url}/users/${this.userId}/live_messages`,
+      `${url}/users/${this.userId}/live_messages` + (
+        this.lastEventId
+          ? `?last_event_id=${this.lastEventId}`
+          : ''
+      ),
```
Comment (grignards) [inline frontend/src/util/LiveMessageManager.js:80]: I think those changes are not part of this issue…

File: frontend/src/container/Header.jsx (hunk 3 above)
Comment (lebouquetin) [reply frontend/src/container/Header.jsx:80]: Singleton is a bad design. The right way to implement this behavior is:
- implement a normal class
- implement a static global shared instance (can be implemented through a factory that manage a unique instance globally)

I'm pretty sure this pattern is used in backend code for an example

Comment (raphj) [reply frontend/src/container/Header.jsx:80]: @lebouquetin a factory has pretty much the same drawbacks as a singleton, of which I am aware (except that it forbids creating a new instance by construction). I had my reasons to go this way, you can't tell me this is bad design without asking why I did this.

You are both right though, it is best to avoid these patterns and after discussion with @grignards we found an elegant solution that solves the problem I am trying to  solve (avoid a bad refresh of the whole app when the connection goes live, which is completely unnecessary and breaks frontend tests), removes the need for a factory or a singleton, and makes the code cleaner and easier to work with: the best of every worlds.

File: frontend/src/util/LiveMessageManager.js (hunk 4 above)
Comment (raphj) [reply frontend/src/util/LiveMessageManager.js]: right, I reverted this.

File: frontend/src/container/Header.jsx [hunk 5]
```diff
@@ -40,7 +40,6 @@ const qs = require('query-string')
 
 export class Header extends React.Component {
   componentDidMount () {
-    this.props.dispatchCustomEvent('TRACIM_HEADER_MOUNTED', {})
```
Comment (Skylsmoi) [inline frontend/src/container/Header.jsx:43]: why are you removing this ? I haven't seen you adding it back somewhere else

File: frontend/src/container/Home.jsx [hunk 6]
```diff
@@ -80,14 +81,14 @@ export class Home extends React.Component {
...
-    if (!(Cookies.get(COOKIE_FRONTEND.HIDE_USERNAME_POPUP) || this.props.user.username)) {
+  maybeShowUsernamePopup () {
+    if (!(this.props.user.username || Cookies.get(COOKIE_FRONTEND.HIDE_USERNAME_POPUP))) {
       this.setState({ usernamePopup: true })
     }
   }
 
   handleClickCloseUsernamePopup = () => {
-    this.setState(prevState => ({ usernamePopup: !prevState.usernamePopup }))
+    this.setState({ usernamePopup: this.state.usernamePopup })
```
Comment (Skylsmoi) [inline frontend/src/container/Home.jsx]: This doesn't seem to do anything. What was wrong with the previous version ?

File: frontend/src/container/Home.jsx [hunk 7]
```diff
@@ -133,17 +134,17 @@ export class Home extends React.Component {
   }
 
   handleClickCheckbox = () => {
-    this.setState(prevState => ({ hidePopupCheckbox: !prevState.hidePopupCheckbox }))
+    this.setState({ hidePopupCheckbox: !this.state.hidePopupCheckbox })
```
Comment (Skylsmoi) [inline frontend/src/container/Home.jsx]: why this change ? This is React bad practice

File: frontend/src/container/Home.jsx [hunk 8]
```diff
@@ -65,7 +66,7 @@ export class Home extends React.Component {
   }
 
   componentDidMount () {
-    this.checkUsername()
+    this.maybeShowUsernamePopup()
```
Comment (gbonaspetti) [inline frontend/src/container/Home.jsx:69]: I don't love that a function starts with maybe, some other possibilities: `setDisplayUsernamePopup` or `setUsernamePopupVisibility`
This comment is just a suggestion, if you prefer your name yet, you can just resolve the conversation.

File: frontend/src/container/Home.jsx (hunk 7 above)
Comment (gbonaspetti) [inline frontend/src/container/Home.jsx:137]: same comment about setState

File: frontend/test/src/LiveMessageManager.spec.js [hunk 9]
```diff
@@ -4,25 +4,45 @@ import {
...
+
+const setStatusOrig = LiveMessageManager.prototype.setStatus
+LiveMessageManager.prototype.setStatus = function (status) {
+  setStatusOrig.call(this, status)
+
+  const resolve = (this.statuscallbacks && this.statuscallbacks.shift())
+  resolve && resolve()
+}
+
+LiveMessageManager.prototype.onstatuschange = function (callback) {
```
Comment (Skylsmoi) [inline frontend/test/src/LiveMessageManager.spec.js:27]: shouldn't you name this function in camel case ?

File: frontend/test/src/LiveMessageManager.spec.js [hunk 10]
```diff
@@ -4,25 +4,45 @@ import {
...
-const waitForStatus = async (manager, status, timeout = 25) => {
-  // NOTE SG 2020-07-03 - have to wait a small amount as manager status changes
-  // can be done via a setTimeout
-  var retries = 10
-  while (manager.status !== status && retries > 0) {
-    await new Promise(resolve => setTimeout(resolve, timeout / 10))
-    retries--
-  }
+const waitForStatus = async (manager) => {
+  await new Promise(resolve => { manager.onstatuschange(resolve) })
```
Comment (Skylsmoi) [inline frontend/test/src/LiveMessageManager.spec.js:45]: I'm not sure I understand the purpose awaiting a new Promise just to give the callback to onstatuschange.
Can you explain the behavior of this please ?

File: frontend/src/container/Home.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (raphj) [reply frontend/src/container/Home.jsx:91]: Interesting, thanks. I am going to revert it.

File: frontend/src/container/Home.jsx (hunk 7 above)
Comment (raphj) [reply frontend/src/container/Home.jsx]: Resolving this, @gbonaspetti already raised this issue.

File: frontend/test/src/LiveMessageManager.spec.js (hunk 9 above)
Comment (raphj) [reply frontend/test/src/LiveMessageManager.spec.js:27]: I figured I would follow the case for the event handlers used everywhere on the DOM. Are you fine with this or should I change this?

(onmousemove, onreadystatechange, etc)

File: frontend/src/container/Header.jsx (hunk 5 above)
Comment (raphj) [reply frontend/src/container/Header.jsx:43]: It's not used anywhere as far as I could see. It's therefore unnecessary and untested code, and I was trying to change unneeded state changes in this PR.

Should I revert this anyway?

File: frontend/test/src/LiveMessageManager.spec.js (hunk 9 above)
Comment (Skylsmoi) [reply frontend/test/src/LiveMessageManager.spec.js:27]: mmh it makes sense but it also surprised me to see all your new functions in camel case but this one.
I'd say to put it in camel case to match all the others functions name of the file

File: frontend/test/src/LiveMessageManager.spec.js (hunk 10 above)
Comment (raphj) [reply frontend/test/src/LiveMessageManager.spec.js:45]: I am waiting for the Promise to resolve. The code continues executing as soon as the promise resolves.

File: frontend/src/container/Header.jsx (hunk 5 above)
Comment (Skylsmoi) [reply frontend/src/container/Header.jsx:43]: It is useful, keep it.
The purpose of this is to allow external apps to do stuffs based on Tracim's event

Comment (raphj) [reply frontend/src/container/Header.jsx:43]: all right! I'm going to revert this.

I have a question however: isn't this prematurilly guessing what will be needed in an uncertain future on a architecture that may change sooner or later anyway?

File: frontend/test/src/LiveMessageManager.spec.js (hunk 10 above)
Comment (Skylsmoi) [reply frontend/test/src/LiveMessageManager.spec.js:45]: shouldn't .onstatuschange() be asynchronous then ? That way you wouldn't have to set a callback mechanism

Comment (raphj) [reply frontend/test/src/LiveMessageManager.spec.js:45]: Mhm, yes it is, and I did. For `onstatuschange`, resolve is just a regular callback here. I'm using it in a Promise, but it does not know about this.

File: frontend/test/src/LiveMessageManager.spec.js (hunk 9 above)
Comment (raphj) [reply frontend/test/src/LiveMessageManager.spec.js:27]: ok :-)

File: frontend/test/src/LiveMessageManager.spec.js (hunk 10 above)
Comment (raphj) [reply frontend/test/src/LiveMessageManager.spec.js]: I turned this function into an asynchronous function and used camel case. Does it looks good to you @Skylsmoi?

File: frontend/src/util/LiveMessageManager.js [hunk 11]
```diff
@@ -9,62 +10,99 @@ export const LIVE_MESSAGE_STATUS = {
...
-      `${url}/users/${userId}/live_messages`,
+    this.setStatus(LIVE_MESSAGE_STATUS.PENDING)
+
+    if (!this.broadcastChannel) {
+      this.broadcastChannel = new BroadcastChannel(BROADCAST_CHANNEL_NAME)
+      this.broadcastChannel.addEventListener('message', this.broadcastChannelMessageReceived.bind(this))
+      this.broadcastChannel.postMessage({ canIHaveStatus: true })
+      this.electLeader()
+    }
+  }
```
[the remark this answers is not in the document]
Comment (raphj) [reply frontend/src/util/LiveMessageManager.js:51]: it is done in `closeLiveMessageConnection` @grignards, or do you speak about something else?

**Review by gbonaspetti — APPROVED** [review]

File: frontend/test/src/LiveMessageManager.spec.js [hunk 12]
```diff
@@ -21,12 +21,12 @@ LiveMessageManager.prototype.setStatus = function (status) {
   setStatusOrig.call(this, status)
 
   const resolve = (this.statuscallbacks && this.statuscallbacks.shift())
-  resolve && resolve()
+  resolve && resolve(this.status)
 }
 
-LiveMessageManager.prototype.onstatuschange = function (callback) {
+LiveMessageManager.prototype.onStatusChange = async function () {
```
Comment (Skylsmoi) [inline frontend/test/src/LiveMessageManager.spec.js:27]: I'm confused because you only `await` for this function in unit tests, I'm not sure to understand

Comment (raphj) [reply frontend/test/src/LiveMessageManager.spec.js:27]: It is only present in unit tests.

Comment (Skylsmoi) [reply frontend/test/src/LiveMessageManager.spec.js:27]: oh, the file is in tests folder ok.

**Review by Skylsmoi — APPROVED** [review]

**Comment by raphj** [comment]
NOT tested on branch, we are going to test this in branch `maintenance/3.0-more_reliable_tlms` with the work on #3409.
